Feat: add Certora specs

amusingaxl opened the pull request.

oddaf committed:
initial implementation

oddaf committed:
Update script/input/1/README.md

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
refactor: rename function put to set

oddaf committed:
fix: add check if data can be safely cast to uint16 in file()

oddaf committed:
refactor: remove step restriction

oddaf committed:
chore: forge fmt

oddaf committed:
refactor: Update Set event to be emitted for each rate update

oddaf committed:
Update src/deployment/DSPCInstance.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCInit.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCInit.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCDeploy.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCDeploy.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCDeploy.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/deployment/DSPCInit.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
refactor: use addresses instead of interfaces in DSPCInstance to ease casting

oddaf committed:
fix: enforce min < max and max > min on file()

oddaf committed:
fix: typo in RelyLike

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
feat: dspc cooldown + step

oddaf committed:
remove unnecessary setup from dspcMom test

oddaf committed:
chore: forge fmt

oddaf committed:
Apply suggestions from code review

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
fix: revert message in test

oddaf committed:
chore: forge fmt

oddaf committed:
feat: improvements to initialization script

oddaf committed:
Apply suggestions from code review

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
chore: adjust tests for new var names

oddaf committed:
chore: improve DSPCInit natspec

oddaf committed:
feat: add bud address to DSPCConfig

oddaf committed:
fix: Remove setAuthority from deploy script in favor of only executing it in the init script

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
refactor: optimize storage layout

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
chore: test case for tau overflow in file()

oddaf committed:
fix: prevent duplicate ilks

oddaf committed:
chore: fix natspec

oddaf committed:
chore: fix readme reference to SUSDS

oddaf committed:
refactor: fisibility of _cfgs to public

oddaf committed:
fix: check if ilk setup on set()

oddaf committed:
chore: Add comment to file function (ilks) mentioned order of calls may impact execution depending on previous values

oddaf committed:
Update src/DSPC.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/DSPC.t.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
fix: update revert reason in test_revert_set_not_configured_rate

oddaf committed:
fix: prevent setup of unexisting Ilks

oddaf committed:
refactor: ordering of ilks check

oddaf committed:
refactor: improve invalid bad check

oddaf committed:
chore: improve natspec dos for file(what, data)

oddaf committed:
refactor: order of checks on set()

oddaf committed:
fix: ilk initialization verification

oddaf committed:
feat: allow ward to edit toc

oddaf committed:
feat: revert if current rate outside range

oddaf committed:
Apply suggestions from code review

Co-authored-by: telome <130504305+telome@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update src/mocks/ConvMock.sol

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
chore: update natspec of set() to include all cases where it reverts

oddaf committed:
chore: improve tests

oddaf committed:
chore: natspec

oddaf committed:
chore: add step to readme.md instructions

oddaf committed:
Update README.md

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Apply suggestions from code review

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
Update README.md

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
chore: order of operations in readme instructions

oddaf committed:
Merge branch 'initial-implementation' of https://github.com/dewiz-xyz/dspc into initial-implementation

oddaf committed:
chore: order natspec revert reasons

oddaf committed:
chore: update foundry.toml

oddaf committed:
chore: formatting

oddaf committed:
chore: remove redundant selectors from test_auth_methods

oddaf committed:
chore: add .editorconfig

oddaf committed:
Apply suggestions from code review

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

oddaf committed:
feat: rate covnersion check

oddaf committed:
chore: forge fmt

oddaf committed:
refactor: allow setting rates when current rate out of bounds

oddaf committed:
chore: test rates below min bound

oddaf committed:
chore: formatting

oddaf committed:
Apply suggestions from code review

Co-authored-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>
Signed-off-by: oddaf <106770775+oddaf@users.noreply.github.com>

amusingaxl committed:
feat: add initial spec [WIP]

amusingaxl committed:
feat: include additional rules and harness

pedrobergamini commented on certora/DSPC.spec:
```diff
@@ -0,0 +1,376 @@
+}
+
+// Verify correct storage changes for non-reverting set for a single rate.
+rule set_single(bytes32 id, uint256 bps) {
```
I've created [this PR](https://github.com/dewiz-xyz/dspc/pull/8) suggesting a `set` rule definition calling `DSPC::set` instead of using the `DSPCHarness`.

amusingaxl committed:
WIP: checkpoint for rules implementation

amusingaxl committed:
fix: mocks and specs to allow rule passing

amusingaxl committed:
refactor: clean up mocks and spec

amusingaxl committed:
feat: add remaining rules for edge cases

amusingaxl committed:
chore: formatting

amusingaxl committed:
refactor: formatting and cleanup

amusingaxl committed:
feat: add Certora CI action

oddaf committed:
Rename module (DSPC -> SPBEAM) (#9)

* refactor: rename module (DSPC -> SPBEAM)

* chore: formatting

oddaf committed:
Rename module (DSPC -> SPBEAM) (#9)

* refactor: rename module (DSPC -> SPBEAM)

* chore: formatting

amusingaxl committed:
fix: remove `MockerBrokenConv` override

The override was failing due to `ConvMock.btor` not being virtual. When I added the `virtual` modifier, for some reason `forge build --sizes` fail because the MockConv contract becomes too big.

amusingaxl committed:
refactor: rename DSPC to SPBEAM

amusingaxl committed:
Merge branch 'initial-implementation' into feat/certora-spec

Signed-off-by: amusingaxl <112016538+amusingaxl@users.noreply.github.com>

amusingaxl committed:
refactor: revert mock implementation change

pedrobergamini commented on .github/workflows/certora.yml:
```diff
@@ -0,0 +1,46 @@
+name: Certora
+
+on: [push, pull_request]
```
maybe trigger the Certora action only on pull requests as it takes some time to run?

pedrobergamini reviewed (dismissed):
Looks good to me! Just commented an idea on Certora's GH actions workflow

amusingaxl committed:
refactor: improve and cleanup rules

pedrobergamini commented on certora/SPBEAM.spec:
```diff
@@ -435,22 +433,21 @@ rule set_revert(SPBEAM.ParamChange[] updates, uint256[] idsAsUints, uint256[] bp
-        revert7 || revert8 || revert9 ||
-        revert10,
-        "set revert rules failed";
+        revert7 || revert8,
```
Nice work on this change

sunbreak1211 commented on certora/SPBEAM.spec:
```diff
@@ -0,0 +1,579 @@
+    env e;
+
+    mathint minBefore; mathint maxBefore; mathint stepBefore;
+    minBefore, maxBefore, stepBefore = cfgs(id);
```
Maybe consider adding min/max/stepOther so for any other id nothing should change. Ofc that is up to how exhaustive the rules want be made, but considering that in others it has been checked might be better to also do here.

sunbreak1211 commented on certora/SPBEAM.spec:
```diff
@@ -0,0 +1,579 @@
+    env e;
+    bytes32 ilk;
+    require ilk != DSR() && ilk != SSR();
+    require updates.length > 0 && updates.length < 4;
```
Could just be `updates.length < 4;` as the == 0 is being a discarded path due to provoke a revert.

amusingaxl replied:
Fixed in b3b2247

sunbreak1211 commented on certora/SPBEAM.spec:
```diff
@@ -0,0 +1,579 @@
+    init_state axiom forall bytes32 i. set_item_reverted[i] == false;
+}
+
+rule set_revert(SPBEAM.ParamChange[] updates, uint256[] idsAsUints, uint256[] bpss) {
```
At least in the current version `bpss` is not fulfilling any purpose.

sunbreak1211 commented on certora/SPBEAM.spec:
```diff
@@ -0,0 +1,579 @@
+        set_item_reverted[updates[2].id] = check_item_revert(e, updates[2].id, updates[2].bps);
+    }
+
+    bool revert8 = exists uint256 i. i < updates.length && set_item_reverted[updates[i].id];
```
`i < updates.length && ` is not needed as all the values are initialized as `false`. It would be needed if initialization is not enforced (which is also an option).

sunbreak1211 replied:
Oh that's weird, I think I have tried it. Anyway nothing really important if working.

amusingaxl commented on certora/SPBEAM.spec:
```diff
@@ -0,0 +1,579 @@
+    bool revert6 = idsAsUints.length == 2 ? idsAsUints[0] >= idsAsUints[1] : false;
+    bool revert7 = idsAsUints.length == 3 ? (
+        idsAsUints[0] >= idsAsUints[1] ||
+        idsAsUints[0] >= idsAsUints[2] ||
```
What line? L422? or the whole `revert7`?

amusingaxl replied:
I guess you're referring to the fact that `a >= b && b >= c => a >= c`, right?

sunbreak1211 replied:
Right, checking that 0 >= 1 and 1 >= 2 is enough for the revert as 0 >= 2 can't be true if the other 2 are false.

sunbreak1211 replied:
And in fact it is what actually the code does, just comparing actual to next value.

amusingaxl replied:
Fixed in f0f2ea9.

amusingaxl committed:
feat(file_per_id): add assertions for keeping other params unchanged

amusingaxl committed:
refactor(set): remove require on updates.length

amusingaxl committed:
refactor(set_revert): apply suggestions from code review

amusingaxl committed:
feat(invariants): add rules check on invariants for the rates being set

amusingaxl commented:
> Maybe an interesting rule/invariant to do, if you end up having some extra time, would be to prove that after any call to set, the values updated can not end up being outside the min-max range and always respect be restricted by the step.

This was added in 67d5fbc.

> Then the fix ordering can be roll back and check that the rule would have caught that.

Not entirely sure if I get what you're suggesting here.

amusingaxl committed:
Merge branch 'master' into feat/certora-spec

amusingaxl committed:
refactor: improve rules and mocks

pedrobergamini commented on certora/mocks/SUsds.sol:
```diff
@@ -1105,19 +188,30 @@ contract SUsds is UUPSUpgradeable {
+        // NOTE: using mocked `drip` function because this method is not a subject of the formal verification
+        // NOTE: this line is needed to prevent `file` above from reverting
+        rho = uint64(block.timestamp);
+        // (uint256 chi_, uint256 rho_) = (chi, rho);
```
Shouldn't we return `chi` here? As right now nChi is always == 0?

```solidity
nChi = uint256(chi)
```

amusingaxl replied:
As discussed, the return value is irrelevant to the main subject of the formal verification.
We just need to ensure that the calls to `drip()` and `file()` in the underlying contracts don't fail for reasons that are not relevant.

pedrobergamini commented on certora/mocks/Jug.sol:
```diff
@@ -175,11 +175,14 @@ contract Jug is LibNote {
+        // NOTE: this line is needed to prevent `file` above from reverting
+        ilks[ilk].rho = now;
+        // require(now >= ilks[ilk].rho, "Jug/invalid-now");
         // (, uint prev) = vat.ilks(ilk);
```
Would it be better if we returned the current `duty` instead of 0 here?

```solidity
function drip(bytes32 ilk) external note returns (uint256 rate) {
    ilks[ilk].rho = now;
    rate = ilks[ilk].duty;  // Return current duty instead of 0
}
```

amusingaxl replied:
Same as above.

pedrobergamini commented on certora/mocks/Pot.sol:
```diff
@@ -204,12 +204,14 @@ contract Pot is LibNote {
         // uint chi_ = sub(tmp, chi);
         // chi = tmp;
-        rho = now;
+        // rho = now;
```
Same question as `Jug::drip`, would it be better if we returned the `dsr` value here?
```solidity
function drip() external note returns (uint256 tmp) {
    rho = now;
    tmp = dsr;  // Return current dsr instead of 0
}
```

amusingaxl replied:
Same as above.

pedrobergamini reviewed (commented):
Added a few questions about the mocks, also, after reviewing the updated rules I thought it could be useful if we added the following one to ensure the config is always consistent:

```cvl
rule config_min_less_than_max(bytes32 id) {
    uint16 min; uint16 max; uint16 step;
    min, max, step = cfgs(id);
    
    assert min <= max, "Configuration min should always be <= max";
}
```

amusingaxl committed:
feat: add rule to check invariant on `min <= max` for the configs

pedrobergamini reviewed (approved):
Looks good to me!
